File: examples/java-kafka-producer/src/main/java/no/njm/WorkshopMessageRestApi.java
package no.njm;

import no.njm.WorkshopMessageProducer.WorkshopMessage;
import no.njm.WorkshopMessageProducer.WorkshopMessageException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.UUID;

@RestController
class WorkshopMessageRestApi {

    private final WorkshopMessageProducer workshopMessageProducer;

    @Autowired
    WorkshopMessageRestApi(WorkshopMessageProducer workshopMessageProducer) {
        this.workshopMessageProducer = workshopMessageProducer;
    }

    @ExceptionHandler(WorkshopMessageException.class)
    ResponseEntity<String> handleException() {
        return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
    }

    @PostMapping("/message")
    ResponseEntity<String> sendMessage(@RequestBody Message message) {
        if (message != null) {
            workshopMessageProducer.sendSync(toWorkshopMessage(message));
            return new ResponseEntity<>(HttpStatus.ACCEPTED);
        }
        return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
    }

    @PostMapping("/message/async")
    ResponseEntity<String> sendAsyncMessage(@RequestBody Message message) {
        if (message != null) {
            workshopMessageProducer.sendAsync(toWorkshopMessage(message));
            return new ResponseEntity<>(HttpStatus.ACCEPTED);
        }
        return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
    }

    private record Message(String message) {
    }

    private WorkshopMessage toWorkshopMessage(Message message) {
        String uuid = UUID.randomUUID().toString();
        return new WorkshopMessage(uuid, message.message);
    }
}

File: examples/kotlin-spring-kafka-producer/src/main/kotlin/no/njm/WorkshopMessageRestApi.kt
package no.njm

import com.fasterxml.jackson.annotation.JsonInclude
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController
class WorkshopMessageRestApi(
    private val workshopMessageProducer: WorkshopMessageProducer
) {

    @ExceptionHandler(WorkshopMessageException::class)
    fun handleException(): ResponseEntity<Any> {
        return ResponseEntity<Any>(HttpStatus.INTERNAL_SERVER_ERROR)
    }

    @PostMapping("/message")
    fun sendMessage(@RequestBody message: Message?): ResponseEntity<Any> {
        message?.let {
            workshopMessageProducer.sendSync(message.toWorkshopMessage())
            return ResponseEntity<Any>(HttpStatus.ACCEPTED)
        }
        return ResponseEntity<Any>(HttpStatus.BAD_REQUEST)
    }

    @PostMapping("/message/async")
    fun sendAsyncMessage(@RequestBody message: Message?): ResponseEntity<Any> {
        message?.let {
            workshopMessageProducer.sendAsync(message.toWorkshopMessage())
            return ResponseEntity<Any>(HttpStatus.ACCEPTED)
        }
        return ResponseEntity<Any>(HttpStatus.BAD_REQUEST)
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    data class Message(
        var message: String,
    )

    private fun Message.toWorkshopMessage(): WorkshopMessage =
        WorkshopMessage(UUID.randomUUID().toString(), message)
}